### ISSUE TITLE
Run tests using UndefinedBehaviorSanitizer

### ISSUE THREAD

**Issue by rcseacord** [zcash/zcash#1318]
The UndefinedBehavior Sanitizer is available for both Clang http://clang.llvm.org/docs/UndefinedBehaviorSanitizer.html and gcc https://gcc.gnu.org/onlinedocs/gcc/Instrumentation-Options.html. 

UndefinedBehaviorSanitizer (UBSan) is a fast undefined behavior detector. UBSan modifies the program at compile-time to catch various kinds of undefined behavior during program execution, for example:
- Using misaligned or null pointer
- Signed integer overflow
- Conversion to, from, or between floating-point  types which would overflow the destination

The following are the changes that I made to instrument the Zcash build to instrument the code using UndefinedBehaviorSanitizer:

```
rcs@rcs-VirtualBox:~/Test/zcashs/zcash$ git diff depends/packages/libsnark.mk
```
```diff
diff --git a/depends/packages/libsnark.mk b/depends/packages/libsnark.mk
index 15912f1..46bf6cd 100644
--- a/depends/packages/libsnark.mk
+++ b/depends/packages/libsnark.mk
@@ -9,7 +9,7 @@ $(package)_git_commit=11242d8afe304c8bcc252baec483f04a110d1dd4
 $(package)_dependencies=libgmp libsodium

 define $(package)_build_cmds
-  CXXFLAGS="-fPIC -DBINARY_OUTPUT -DNO_PT_COMPRESSION=1" $(MAKE) lib DEPINST=$(host_prefix) CURVE=ALT_BN128 MULTICORE=1 NO_PROCPS=1 NO_GTEST=1 NO_DOCS=1 STATIC=1 NO_SUPERCOP=1 FEATUREFLAGS=-DMONTGOMERY_OUTPUT
-  CXXFLAGS="-static-libstdc++ -static-libubsan -fPIC -DBINARY_OUTPUT -DNO_PT_COMPRESSION=1 -Wall -Wextra -Wno-unused-parameter -g -Wno-comment -Wfatal-errors -Waddress -Wformat-signedness -Wsequence-point -Wmaybe-uninitialized -Wstrict-overflow=3 -Wpointer-arith -Wcast-align -Wconversion -Wsign-compare -Wsign-conversion -Wfloat-conversion -Wmemset-transposed-args -Wno-aggressive-loop-optimizations -O2 -fsanitize=undefined" $(MAKE) lib DEPINST=$(host_prefix) CURVE=ALT_BN128 MULTICORE=1 NO_PROCPS=1 NO_GTEST=1 NO_DOCS=1 STATIC=1 NO_SUPERCOP=1 FEATUREFLAGS=-DMONTGOMERY_OUTPUT
  endef
  
  define $(package)_stage_cmds
```
```
  rcs@rcs-VirtualBox:~/Test/zcashs/zcash$ git diff src/Makefile.am
```
```diff
diff --git a/src/Makefile.am b/src/Makefile.am
index 7c843cb..3b628e0 100644
--- a/src/Makefile.am
+++ b/src/Makefile.am
@@ -1,4 +1,4 @@
-DIST_SUBDIRS = secp256k1
+DPIST_SUBDIRS = secp256k1
 AM_LDFLAGS = $(PTHREAD_CFLAGS) $(LIBTOOL_LDFLAGS)

@@ -415,7 +415,7 @@ libzcash_a_SOURCES = \
   zcash/prf.cpp \
   zcash/util.cpp

-libzcash_a_CPPFLAGS = -DMULTICORE -fopenmp -fPIC -DBINARY_OUTPUT -DCURVE_ALT_BN128 -DBOOST_SPIRIT_THREADSAFE -DHAVE_BUILD_INFO -D__STDC_FORMAT_MACROS $(HARDENED_CPPFLAGS) $(HARDENED_CXXFLAGS) $(HARDENED_LDFLAGS) -pipe -O1 -g -Wstack-protector -fstack-protector-all -fPIE -fvisibility=hidden -DSTATIC $(BITCOIN_INCLUDES)
+libzcash_a_CPPFLAGS = -DMULTICORE -fopenmp -static-libstdc++ -static-libubsan -fPIC -DBINARY_OUTPUT -DCURVE_ALT_BN128 -DBOOST_SPIRIT_THREADSAFE -DHAVE_BUILD_INFO -D__STDC_FORMAT_MACROS $(HARDENED_CPPFLAGS) $(HARDENED_CXXFLAGS) $(HARDENED_LDFLAGS) -pipe -O2 -g -Wstrict-overflow=3 -Wstack-protector -fsanitize=undefined -fomit-frame-pointer -fstack-protector-all -fvisibility=hidden -DSTATIC $(BITCOIN_INCLUDES)

 # bitcoinconsensus library #
 if BUILD_BITCOIN_LIBS
```
```
rcs@rcs-VirtualBox:~/Test/zcashs/zcash$ git diff zcutil/build.sh
```
```diff
diff --git a/zcutil/build.sh b/zcutil/build.sh
index 6f79157..6c5b4c6 100755
--- a/zcutil/build.sh
+++ b/zcutil/build.sh
@@ -35,5 +35,5 @@ PREFIX="$(pwd)/depends/x86_64-unknown-linux-gnu/"

 make "$@" -C ./depends/ V=1 NO_QT=1
 ./autogen.sh
-./configure --prefix="${PREFIX}" --with-gui=no --enable-hardening "$LCOV_ARG" CXXFLAGS='-Wno-deprecated-declarations -Wno-placement-new -Wno-terminate -Werror -O1 -g'
+./configure --prefix="${PREFIX}" --with-gui=no --enable-hardening "$LCOV_ARG" CXXFLAGS='-Wno-deprecated-declarations -Wno-placement-new -Wno-terminate -Werror -O2 -static-libstdc++ -static-libubsan -g -fsanitize=undefined'
 make "$@" V=1
```

**Comment by nathan-at-least** [comment]
Bumped, due to time constraints. Notice the history of postponement here! This is evidence of project coordination problems.

**Comment by mdr0id** [comment]
Closed by #3048
